_posts/2017-04-12-managing-technical-debt.md

```diff
@@ -0,0 +1,83 @@
+13. **Create template for new issues/bugs/ stories in the issue tracker.** Without good descriptions in your issues they became an absolutely useless. This is necessary bureaucracy to save developers time and focus. 
+14. **Investigate most confusing and affective parts in application** and create a plan to how to make them better
+15. **Start creating tests for all newly created code**, and insist on them on code reviews (at least acceptance tests should cover the issue).
+16. **Follow the “The Boy Scouts Rule”**, create the tests for part you touch and refactor it. Try to spend on this at least 20% of time.  
+
+# LR: Details
+So, where is the technical debt most often accumulated? On a new project team usually:
+
+1. Fear to make the changes
+2. Has no completely understanding of the application domain logic 
+3. Skip unit/functional/acceptance/performance tests, and testing infrastructure at all
+4. Implement manual deployment strategy and infrastructure configuration
+5. Do not track requirements
+6. Configure Dev environment and onboarding manually
+7. Has no backup automation for data
+8. Has no data sampling for dev/QA purposes 
+9. Has a lot of “code smells” in source code, design by committee with a lot of hidden requirements due to lack of refactoring
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe each a little. 
+
+From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramastically with no chance to become better. 
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users. 
+
+Without project requirements management impossible to create good deployment and onboarding process, so both can take a days or, sometimes, weeks. 
+
+Without useful developer environment, developers will spend a lot of time trying to figure out is some particular issue are related to their local environment, or it’s a general issue; some thing can’t be run or tested on a local machine at all (for example delayed jobs, file storing in case when it’s configured to store on the third-party resources like S3, mail sending etc.); will be not able to reproduce bugs from other developers or production; it will be hard to implement configuration or new requirements without  breaking other team member processes; 
```

Comment:
Without useful developer environment, developers will spend a lot of time trying to figure out [ if the particular issue is related to their local environment]